tests/autocorrect/test_autocorrect.cpp

```diff
@@ -0,0 +1,88 @@
+/* Copyright 2017 Fred Sundvik
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "keycode.h"
+#include "test_common.hpp"
+
+using testing::_;
+using testing::InSequence;
```

Comment:
It seems you intended to use `InSequence` below to ensure the order of expected calls. Particularly, this would constrain the order of the "backspace, S, E" reports instead of allowing them in any order.

https://github.com/google/googletest/blob/main/docs/gmock_for_dummies.md#ordered-vs-unordered-calls-orderedcalls

Reply:
To be honest, I have almost no idea what I'm doing with the tests.  Any help and guidance with them is greatly appreciated, as I'm actively learning how to implement them.  (this PR was the motivation for that, as well).

Reply:
No worries! gmock is a rather confusing API, I recently went through trying to learn it myself for Caps Word. You can check out the test here:
https://github.com/qmk/qmk_firmware/blob/9fb91771b5920011338b0a1f42d23a1e1cb87ee4/tests/caps_word/test_caps_word.cpp

Here, I think something like this would do it: 

```c++
// Allow any number of empty reports.
EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport())).Times(AnyNumber());
{  // Expect the following reports in this order.
    InSequence s;
EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_BACKSPACE)));
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_S)));
    EXPECT_CALL(driver, send_keyboard_mock(KeyboardReport(KC_E)));
}
```

Reply:
Thanks!  That actually helps a lot, and has helped me to better understand what is going on. 

And there are a few more tests I'd like to add, just to make the test much more robust, and catch more potential edge cases.

Reply:
Well done! That's a solid suite of tests.

> And yeah, I learn best by example. Eg, if I can see what is going on, it helps me understand it much better.

Same here. Looking at this PR was extremely useful in putting together my PR for Caps Word.

> on a side note: tests on macOS are a PITA, since xcode's gcc compiler is very much out of date.

Is XCode GCC stuck on an old version of the C++ standard? I assume it's at least C++11, since googletest needs that.

Reply:
Thanks!  Figured it would be a good idea to test as many scenarios as possible, just in case.   

Yeah, the xcode gcc is stuck on version 4.2.1, with some additional config (clang support, etc).  It's a mess. Even setting gcc with the bash/zsh rc doesn't get picked up correctly. But for anyone interested, `make test:(name) CC=/usr/local/bin/gcc-11` works to run it locally.